File: Backup/SVL 3/ESystem/_Extensions/ObjectExt.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Collections;
using System.Reflection;

namespace ESystem.Extensions
{
  public static class ObjectExt
  {
    private const string NULL = "(null)";

    #region "InLine"

    private static int callIndex = 0;
    public static string ToInlineInfoString(this object obj)
    {
      callIndex += 1;
      if (callIndex > 100)
        Console.WriteLine("stop");

      if ((obj == null))
        return NULL;
      else if (obj is string)
        return Convert.ToString(obj);

      StringBuilder cont = new StringBuilder();
      System.Text.StringBuilder ret = new System.Text.StringBuilder();
      Type t = obj.GetType();

      if (obj is ICollection)
      {
        ICollection ie = (ICollection)obj;

        foreach (object fItem in ie)
        {
          string valueString = ToInlineInfoString(fItem);
          cont.AppendPreDelimited(valueString, ';');
        }
      }
      else if (obj is IEnumerable)
      {
        IEnumerable ie = (IEnumerable)obj;

        foreach (object fItem in ie)
        {
          string valueString = ToInlineInfoString(fItem);
          cont.AppendPreDelimited(valueString, ';');
        }
      }
      else
      {
        PropertyInfo[] props = t.GetProperties();
        MemberInfo[] mems = props.ToArray();
        string loc = GenerateInlineContent(obj, mems);

        if ((loc.Length > 0))
          cont.AppendPreDelimited(loc.ToString(), ';');
      }

      if ((cont.Length == 0) && (obj is ICollection == false))
        ret.Append(obj.ToString());
      else
      {
        ret.Append("{" + cont.ToString() + "}");
        ret.Insert(0, "(" + t.Name + ")");
      }

      return ret.ToString();
    }

    private static string GenerateInlineContent(object obj, MemberInfo[] members)
    {
      System.Text.StringBuilder sb = new System.Text.StringBuilder();

      foreach (MemberInfo Item in members)
      {
        
        //object fieldValue = obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, null, obj, null, null);
        object fieldValue = obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, null, obj, null);
        string fieldValueString = ToInlineInfoString(fieldValue);

        sb.AppendPreDelimited(Item.Name + "=" + fieldValueString, ',');
      }

      return sb.ToString();
    }
    #endregion


    #region "MultiLine"

    public static string ToMultilineInfoString(this object obj)
    {
      return ToMultilineInfoString(obj, 0);
    }


    private static string ToMultilineInfoString(object obj, int level)
    {
      string prefix = "  ".Repeat(level);

      if ((obj == null))
        return prefix + NULL;
      else if (obj is string)
        return prefix + Convert.ToString(obj);

      StringBuilder cont = new StringBuilder();
      System.Text.StringBuilder ret = new System.Text.StringBuilder();
      Type t = obj.GetType();

      if (obj is ICollection)
      {
        ICollection ie = (ICollection)obj;

        foreach (object fItem in ie)
        {
          string valueString = ToMultilineInfoString(fItem, level + 1);
          cont.AppendPreDelimited(prefix + "  " + valueString.TrimStart(), Environment.NewLine);
        }
      }
      else if (obj is IEnumerable)
      {
        IEnumerable ie = (IEnumerable)obj;

        foreach (object fItem in ie)
        {
          string valueString = ToMultilineInfoString(fItem, level + 1);
          cont.AppendPreDelimited(prefix + "  " + valueString.TrimStart(), Environment.NewLine);
        }
      }
      else
      {
        PropertyInfo[] props = t.GetProperties();
        MemberInfo[] mems = props.ToArray();

        string loc = GenerateMultilineContent(obj, mems, level);

        if ((loc.Length > 0))
        {
          cont.AppendPreDelimited(loc.ToString(), ';');
        }
      }

      if ((cont.Length == 0))
      {
        ret.Append(prefix + obj.ToString());
      }
      else
      {
        ret.Append(Environment.NewLine + cont.ToString());

        ret.Insert(0, prefix + "(" + t.Name + ")");
      }

      return ret.ToString();
    }

    private static string GenerateMultilineContent(object obj, MemberInfo[] members, int level)
    {
      System.Text.StringBuilder sb = new System.Text.StringBuilder();

      string prefix = "  ".Repeat(level) + " > ";

      foreach (MemberInfo Item in members)
      {
        // object fieldValue = obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, null, obj, null, null);
        object fieldValue = obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, null, obj, null);
        string fieldValueString = ToMultilineInfoString(fieldValue, level + 1);

        sb.AppendPreDelimited(prefix + Item.Name + "=" + fieldValueString.TrimStart(), Environment.NewLine);
      }

      return sb.ToString();
    }

    #endregion


    #region "XML"

    #region "Nested"

    private const string INTENDER = "  ";
    private const string DEFAULT_ENCAPSULATING_TAG = "This";
    private const string DEFAULT_ENUMERATIONITEM_TAG = "Item";
    private const string DEFAULT_NULL_TAGITEM = "(null)";
    public class Settings
    {
      internal bool useLongTypeNames = false;
      internal bool expandValueTypes = false;
      internal string encapsulatingTag = ObjectExt.DEFAULT_ENCAPSULATING_TAG;
      internal string enumerationItemTag = ObjectExt.DEFAULT_ENUMERATIONITEM_TAG;
      internal int maxDepth = 3;
      internal int currentDepth = 0;
      internal bool sortByName = false;
      internal bool includeFields = true;
      internal bool includeProperties = true;
      internal bool expandEnumerations = true;
      internal string nullTagItem = ObjectExt.DEFAULT_NULL_TAGITEM;
      internal bool inlineValues = false;
    }

    private class MemberInfoByNameComparer : IComparer<MemberInfo>
    {

      public int Compare(System.Reflection.MemberInfo x, System.Reflection.MemberInfo y)
      {
        return x.Name.CompareTo(y.Name);
      }

    }

    #endregion


    #region "Public methods"

    public static string ToXml(this object item)
    {
      string ret = _ToXml(item, new Settings());
      return ret;
    }

    public static string ToXml(this object item, bool useLongTypeNames, string encapsulatingTag)
    {
      Settings sett = new Settings();
      sett.useLongTypeNames = useLongTypeNames;
      sett.encapsulatingTag = encapsulatingTag;

      string ret = _ToXml(item, sett);

      return ret;

    }

    public static string ToXml(this object item, bool useLongTypeNames, string encapsulatingTag, bool inlineValues)
    {
      Settings sett = new Settings();
      sett.useLongTypeNames = useLongTypeNames;
      sett.encapsulatingTag = encapsulatingTag;
      sett.inlineValues = inlineValues;

      string ret = _ToXml(item, sett);

      return ret;

    }

    public static string ToXml(this object item, 
      bool useLongTypeNames, bool expandValueTypes, 
      bool includeProperties, bool includeFields, 
      bool expandEnumerations, bool sortByName)
    {
      Settings sett = new Settings();
      sett.useLongTypeNames = useLongTypeNames;
      sett.expandValueTypes = expandValueTypes;
      sett.sortByName = sortByName;
      sett.includeFields = includeFields;
      sett.includeProperties = includeProperties;
      sett.expandEnumerations = expandEnumerations;

      string ret = _ToXml(item, sett);

      return ret;

    }

    public static string ToXml(this object item, 
      bool useLongTypeNames, bool expandValueTypes, 
      bool includeProperties, bool includeFields, 
      bool expandEnumerations, bool sortByName, int maxDepth)
    {
      Settings sett = new Settings();
      sett.useLongTypeNames = useLongTypeNames;
      sett.expandValueTypes = expandValueTypes;
      sett.maxDepth = maxDepth;
      sett.sortByName = sortByName;
      sett.includeFields = includeFields;
      sett.includeProperties = includeProperties;
      sett.expandEnumerations = expandEnumerations;

      string ret = _ToXml(item, sett);

      return ret;

    }

    public static string ToXml(this object item, 
      bool useLongTypeNames, bool expandValueTypes, 
      bool includeProperties, bool includeFields, 
      bool expandEnumerations, bool sortByName, 
      string encapsulatingTag, string enumerationItemTag, string nullTagItem,
      int maxDepth, bool inlineValues)
    {
      Settings sett = new Settings();
      sett.useLongTypeNames = useLongTypeNames;
      sett.expandValueTypes = expandValueTypes;
      sett.encapsulatingTag = encapsulatingTag;
      sett.enumerationItemTag = enumerationItemTag;
      sett.maxDepth = maxDepth;
      sett.sortByName = sortByName;
      sett.includeFields = includeFields;
      sett.includeProperties = includeProperties;
      sett.expandEnumerations = expandEnumerations;
      sett.nullTagItem = nullTagItem;
      sett.inlineValues = inlineValues;

      string ret = _ToXml(item, sett);

      return ret;

    }

    #endregion

    #region "Private methods"

    private static string _ToXml(object item, Settings sett)
    {
      if (item == null)
      {
        throw new ArgumentNullException("Cannot generate XML content from null/nothing value.");
      }

      string ret = CreateXmlByType(sett.encapsulatingTag, item, sett);

      ret = FormatXml(ret);

      return ret;

    }

    private static string CreateXml(string tag, ValueType value, Settings sett)
    {
      string ret = CreateTag(tag, value.GetType(), true, value.ToString(), sett);

      return ret;
    }

    private static string CreateXml(string tag, string value, Settings sett)
    {
      string ret = CreateTag(tag, value.GetType(), true, value, sett);

      return ret;
    }

    private static string CreateXml(string tag, Enum value, Settings sett)
    {
      string ret = CreateTag(tag, value.GetType(), true, value.ToString(), sett);

      return ret;
    }

    private static string CreateXml(string tag, IEnumerable value, Settings sett)
    {
      StringBuilder content = new StringBuilder();
      string pom = null;


      if (sett.expandEnumerations)
      {
        foreach (object fItem in value)
        {
          pom = CreateXmlByType(sett.enumerationItemTag, fItem, sett);
          content.AppendLine(pom);
        }

      }
      else
      {
        if ((value is ICollection))
        {
          content.AppendLine("Count: " + ((ICollection)value).Count.ToString());
        }
        else
        {
          content.AppendLine();
        }
      }

      string ret = CreateTag(tag, value.GetType(), false, content.ToString(), sett);

      return ret;

    }


    private static string CreateXml(string tag, object value, Settings sett)
    {
      StringBuilder content = new StringBuilder();
      string pom = null;
      object itemValue = null;
      Type valueType = value.GetType();
      PropertyInfo[] props = valueType.GetProperties();
      FieldInfo[] fields = valueType.GetFields();
      List<MemberInfo> members = new List<MemberInfo>();

      if (sett.includeFields)
        members.AddRange(valueType.GetFields());
      if (sett.includeProperties)
        members.AddRange(valueType.GetProperties());

      if (sett.sortByName)
      {
        members.Sort(new MemberInfoByNameComparer());
      }


      foreach (MemberInfo fItem in members)
      {

        if (fItem is PropertyInfo)
        {
          PropertyInfo pi = (PropertyInfo)fItem;
          if (!pi.CanRead)
            continue;

          //itemValue = valueType.InvokeMember(pi.Name, BindingFlags.GetProperty, null, value, null, null);
          itemValue = valueType.InvokeMember(pi.Name, BindingFlags.GetProperty, null, value, null);
        }
        else
        {
          //itemValue = valueType.InvokeMember(fItem.Name, BindingFlags.GetField, null, value, null, null);
          itemValue = valueType.InvokeMember(fItem.Name, BindingFlags.GetField, null, value, null);
        }

        if (itemValue == null)
        {
          Type t = null;
          if (fItem is PropertyInfo)
          {
            t = ((PropertyInfo)fItem).PropertyType;
          }
          else
          {
            t = ((FieldInfo)fItem).FieldType;
          }

          pom = CreateTag(fItem.Name, t, true, sett.nullTagItem, sett);
          content.AppendLine(pom);

        }
        else
        {
          pom = CreateXmlByType(fItem.Name, itemValue, sett);
          content.AppendLine(pom);
        }

      }

      string ret = CreateTag(tag, value.GetType(), false, content.ToString(), sett);

      return ret;

    }


    private static string CreateXmlByType(string tag, object value, Settings sett)
    {
      if (sett.maxDepth < sett.currentDepth)
        return "";

      sett.currentDepth += 1;

      string ret = null;

      if (value is ValueType)
      {
        Type t = value.GetType();
        if (t.IsPrimitive)
        {
          ret = CreateXml(tag, (ValueType)value, sett);
        }
        else
        {
          if (sett.expandValueTypes)
          {
            ret = CreateXml(tag, (object)value, sett);
          }
          else
          {
            ret = CreateXml(tag, (ValueType)value, sett);
          }
        }
      }
      else if (value is string)
      {
        ret = CreateXml(tag, Convert.ToString(value), sett);
      }
      else if (value is IEnumerable)
      {
        ret = CreateXml(tag, (IEnumerable)value, sett);
      }
      else
      {
        ret = CreateXml(tag, (object)value, sett);
      }

      sett.currentDepth -= 1;

      return ret;
    }

    private delegate StringBuilder AddingDelegate(string text);

    private static string CreateTag(string tag, Type type, bool isAtomic, string value, Settings sett)
    {
      StringBuilder sb = new StringBuilder();
      AddingDelegate d = null;

      if (sett.inlineValues && isAtomic)
      {
        d = sb.Append;
      }
      else
      {
        d = sb.AppendLine;
      }

      if (sett.useLongTypeNames)
      {
        d("<" + tag + " type=\"" + type.FullName + "\">");
      }
      else
      {
        d("<" + tag + " type=\"" + type.Name + "\">");
      }

      d(value);

      d("</" + tag + ">");

      return sb.ToString();
    }

    private static string FormatXml(string text)
    {
      System.IO.StringReader rdr = new System.IO.StringReader(text);
      int level = 0;
      StringBuilder ret = new StringBuilder();

      string line = rdr.ReadLine();

      while (line != null)
      {
        if (line.StartsWith("</"))
        {
          level = level - 1;
          ret.AppendLine(INTENDER.Repeat(level) + line);
        }
        else if (line.StartsWith("<"))
        {
          ret.AppendLine(INTENDER.Repeat(level) + line);
          if ((!line.Contains("</")))
            level = level + 1;
        }
        else if (line.Trim().Length > 0)
        {
          ret.AppendLine(INTENDER.Repeat(level) + line);
        }

        line = rdr.ReadLine();
      }

      return ret.ToString();
    }

    #endregion

    #endregion
  }

}

File: FW 4.0/ESystem/_Extensions/ObjectExt.vb
Option Explicit On
Option Strict On

Imports System.Runtime.CompilerServices
Imports System.Reflection
Imports System.Text

Namespace Extensions

  Public Module ObjectExt

    Private Const NULL As String = "(null)"


#Region "InLine"

    <Extension()> _
    Public Function ToInlineInfoString(ByVal obj As Object) As String

      If (obj Is Nothing) Then
        Return NULL
      ElseIf TypeOf obj Is String Then
        Return CType(obj, String)
      End If

      Dim cont As New StringBuilder()
      Dim ret As New System.Text.StringBuilder()
      Dim t As Type = obj.GetType()

      If TypeOf obj Is ICollection Then
        Dim ie As ICollection = CType(obj, ICollection)

        For Each fItem As Object In ie
          Dim valueString As String = _
            ToInlineInfoString(fItem)

          cont.AppendPreDelimited(valueString, ";"c)

        Next

      ElseIf TypeOf obj Is IEnumerable Then
        Dim ie As IEnumerable = CType(obj, IEnumerable)

        For Each fItem As Object In ie
          Dim valueString As String = _
            ToInlineInfoString(fItem)

          cont.AppendPreDelimited(valueString, ";"c)

        Next
      Else

        Dim props() As PropertyInfo = t.GetProperties()
        Dim mems() As MemberInfo = props.ToArray()

        Dim loc As String = GenerateInlineContent( _
          obj, mems)

        If (loc.Length > 0) Then
          cont.AppendPreDelimited(loc.ToString(), ";"c)
        End If

      End If

      If (cont.Length = 0) Then
        ret.Append(obj.ToString())
      Else
        ret.Append("{" + cont.ToString() + "}")

        ret.Insert(0, "(" + t.Name + ")")
      End If

      Return ret.ToString()
    End Function

    Private Function GenerateInlineContent( _
      ByVal obj As Object, ByVal members() As MemberInfo) As String
      Dim sb As New System.Text.StringBuilder()

      For Each Item As MemberInfo In members
        Dim fieldValue As Object = _
          obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, Nothing, obj, Nothing, Nothing)
        Dim fieldValueString As String = _
          ToInlineInfoString(fieldValue)

        sb.AppendPreDelimited(Item.Name + "=" + fieldValueString, ","c)
      Next

      Return sb.ToString()
    End Function
#End Region 'InLine


#Region "MultiLine"

    <Extension()> _
       Public Function ToMultilineInfoString(ByVal obj As Object) As String

      Return ToMultilineInfoString(obj, 0)
    End Function

    Private Function ToMultilineInfoString(ByVal obj As Object, ByVal level As Integer) As String

      Dim prefix As String = "  ".Repeat(level)

      If (obj Is Nothing) Then
        Return prefix & NULL
      ElseIf TypeOf obj Is String Then
        Return prefix & CType(obj, String)
      End If

      Dim cont As New StringBuilder()
      Dim ret As New System.Text.StringBuilder()
      Dim t As Type = obj.GetType()

      If TypeOf obj Is ICollection Then
        Dim ie As ICollection = CType(obj, ICollection)

        For Each fItem As Object In ie
          Dim valueString As String = _
            ToMultilineInfoString(fItem, level + 1)

          cont.AppendPreDelimited(prefix & "  " & valueString.TrimStart(), vbCrLf)

        Next

      ElseIf TypeOf obj Is IEnumerable Then
        Dim ie As IEnumerable = CType(obj, IEnumerable)

        For Each fItem As Object In ie
          Dim valueString As String = _
            ToMultilineInfoString(fItem, level + 1)

          cont.AppendPreDelimited(prefix & "  " & valueString.TrimStart(), vbCrLf)

        Next
      Else

        Dim props() As PropertyInfo = t.GetProperties()
        Dim mems() As MemberInfo = props.ToArray()

        Dim loc As String = GenerateMultilineContent( _
          obj, mems, level)

        If (loc.Length > 0) Then
          cont.AppendPreDelimited(loc.ToString(), ";"c)
        End If

      End If

      If (cont.Length = 0) Then
        ret.Append(prefix & obj.ToString())
      Else
        ret.Append(vbCrLf & cont.ToString())

        ret.Insert(0, prefix & "(" + t.Name + ")")
      End If

      Return ret.ToString()
    End Function

    Private Function GenerateMultilineContent( _
      ByVal obj As Object, ByVal members() As MemberInfo, ByVal level As Integer) As String
      Dim sb As New System.Text.StringBuilder()

      Dim prefix As String = "  ".Repeat(level) & " > "

      For Each Item As MemberInfo In members
        Dim fieldValue As Object = _
          obj.GetType().InvokeMember(Item.Name, BindingFlags.GetProperty, Nothing, obj, Nothing, Nothing)
        Dim fieldValueString As String = _
          ToMultilineInfoString(fieldValue, level + 1)

        sb.AppendPreDelimited(prefix & Item.Name + "=" + fieldValueString.TrimStart(), vbCrLf)
      Next

      Return sb.ToString()
    End Function

#End Region 'MultiLine


#Region "XML"

#Region "Nested"

    Private Const INTENDER As String = "  "
    Private Const DEFAULT_ENCAPSULATING_TAG As String = "This"
    Private Const DEFAULT_ENUMERATIONITEM_TAG As String = "Item"
    Private Const DEFAULT_NULL_TAGITEM As String = "(null)"
    Public Class Settings
      Friend useLongTypeNames As Boolean = False
      Friend expandValueTypes As Boolean = False
      Friend encapsulatingTag As String = DEFAULT_ENCAPSULATING_TAG
      Friend enumerationItemTag As String = DEFAULT_ENUMERATIONITEM_TAG
      Friend maxDepth As Integer = 3
      Friend currentDepth As Integer = 0
      Friend sortByName As Boolean = False
      Friend includeFields As Boolean = True
      Friend includeProperties As Boolean = True
      Friend expandEnumerations As Boolean = True
      Friend nullTagItem As String = DEFAULT_NULL_TAGITEM
      Friend inlineValues As Boolean = False
    End Class

    Private Class MemberInfoByNameComparer
      Implements IComparer(Of MemberInfo)

      Public Function Compare(ByVal x As System.Reflection.MemberInfo, ByVal y As System.Reflection.MemberInfo) As Integer Implements System.Collections.Generic.IComparer(Of System.Reflection.MemberInfo).Compare
        Return x.Name.CompareTo(y.Name)
      End Function

    End Class

#End Region 'Nested


#Region "Public methods"

    <Extension()> _
  Public Function ToXml(ByVal item As Object) As String
      Dim ret As String = _ToXml(item, New Settings())
      Return ret
    End Function

    <Extension()> _
 Public Function ToXml( _
   ByVal item As Object, _
    ByVal useLongTypeNames As Boolean, _
    ByVal encapsulatingTag As String) As String

      Dim sett As New Settings()
      sett.useLongTypeNames = useLongTypeNames
      sett.encapsulatingTag = encapsulatingTag

      Dim ret As String = _ToXml(item, sett)

      Return ret

    End Function

    <Extension()> _
Public Function ToXml( _
ByVal item As Object, _
ByVal useLongTypeNames As Boolean, _
ByVal encapsulatingTag As String, _
ByVal inlineValues As Boolean) As String

      Dim sett As New Settings()
      sett.useLongTypeNames = useLongTypeNames
      sett.encapsulatingTag = encapsulatingTag
      sett.inlineValues = inlineValues

      Dim ret As String = _ToXml(item, sett)

      Return ret

    End Function

    <Extension()> _
  Public Function ToXml( _
    ByVal item As Object, _
     ByVal useLongTypeNames As Boolean, _
     ByVal expandValueTypes As Boolean, _
     ByVal includeProperties As Boolean, _
     ByVal includeFields As Boolean, _
     ByVal expandEnumerations As Boolean, _
     ByVal sortByName As Boolean _
     ) As String

      Dim sett As New Settings()
      sett.useLongTypeNames = useLongTypeNames
      sett.expandValueTypes = expandValueTypes
      sett.sortByName = sortByName
      sett.includeFields = includeFields
      sett.includeProperties = includeProperties
      sett.expandEnumerations = expandEnumerations

      Dim ret As String = _ToXml(item, sett)

      Return ret

    End Function

    <Extension()> _
  Public Function ToXml( _
    ByVal item As Object, _
     ByVal useLongTypeNames As Boolean, _
     ByVal expandValueTypes As Boolean, _
     ByVal includeProperties As Boolean, _
     ByVal includeFields As Boolean, _
     ByVal expandEnumerations As Boolean, _
     ByVal sortByName As Boolean, _
     ByVal maxDepth As Integer) As String

      Dim sett As New Settings()
      sett.useLongTypeNames = useLongTypeNames
      sett.expandValueTypes = expandValueTypes
      sett.maxDepth = maxDepth
      sett.sortByName = sortByName
      sett.includeFields = includeFields
      sett.includeProperties = includeProperties
      sett.expandEnumerations = expandEnumerations

      Dim ret As String = _ToXml(item, sett)

      Return ret

    End Function

    <Extension()> _
  Public Function ToXml( _
    ByVal item As Object, _
     ByVal useLongTypeNames As Boolean, _
     ByVal expandValueTypes As Boolean, _
     ByVal includeProperties As Boolean, _
     ByVal includeFields As Boolean, _
     ByVal expandEnumerations As Boolean, _
     ByVal sortByName As Boolean, _
     ByVal encapsulatingTag As String, _
     ByVal enumerationItemTag As String, _
     ByVal nullTagItem As String, _
     ByVal maxDepth As Integer, _
     ByVal inlineValues As Boolean) As String

      Dim sett As New Settings()
      sett.useLongTypeNames = useLongTypeNames
      sett.expandValueTypes = expandValueTypes
      sett.encapsulatingTag = encapsulatingTag
      sett.enumerationItemTag = enumerationItemTag
      sett.maxDepth = maxDepth
      sett.sortByName = sortByName
      sett.includeFields = includeFields
      sett.includeProperties = includeProperties
      sett.expandEnumerations = expandEnumerations
      sett.nullTagItem = nullTagItem
      sett.inlineValues = inlineValues

      Dim ret As String = _ToXml(item, sett)

      Return ret

    End Function

#End Region 'Public methods


#Region "Private methods"

    Private Function _ToXml(ByVal item As Object, ByVal sett As Settings) As String

      If item Is Nothing Then
        Throw New ArgumentNullException("Cannot generate XML content from null/nothing value.")
      End If

      Dim ret As String = CreateXmlByType(sett.encapsulatingTag, item, sett)

      ret = FormatXml(ret)

      Return ret

    End Function

    Private Function CreateXml(ByVal tag As String, ByVal value As ValueType, ByVal sett As Settings) As String
      Dim ret As String = _
        CreateTag(tag, value.GetType(), True, value.ToString(), sett)

      Return ret
    End Function

    Private Function CreateXml(ByVal tag As String, ByVal value As String, ByVal sett As Settings) As String
      Dim ret As String = _
        CreateTag(tag, value.GetType(), True, value, sett)

      Return ret
    End Function

    Private Function CreateXml(ByVal tag As String, ByVal value As [Enum], ByVal sett As Settings) As String
      Dim ret As String = _
        CreateTag(tag, value.GetType(), True, value.ToString(), sett)

      Return ret
    End Function

    Private Function CreateXml(ByVal tag As String, ByVal value As IEnumerable, ByVal sett As Settings) As String
      Dim content As New StringBuilder
      Dim pom As String

      If sett.expandEnumerations Then

        For Each fItem As Object In value
          pom = CreateXmlByType(sett.enumerationItemTag, fItem, sett)
          content.AppendLine(pom)
        Next
      Else

        If (TypeOf value Is ICollection) Then
          content.AppendLine("Count: " + CType(value, ICollection).Count.ToString())
        Else
          content.AppendLine()
        End If
      End If

      Dim ret As String = _
        CreateTag(tag, value.GetType(), False, content.ToString(), sett)

      Return ret

    End Function

    Private Function CreateXml(ByVal tag As String, ByVal value As Object, ByVal sett As Settings) As String

      Dim content As New StringBuilder
      Dim pom As String
      Dim itemValue As Object
      Dim valueType As Type = value.GetType()
      Dim props() As PropertyInfo = valueType.GetProperties()
      Dim fields() As FieldInfo = valueType.GetFields()
      Dim members As New List(Of MemberInfo)

      If sett.includeFields Then _
        members.AddRange(valueType.GetFields())
      If sett.includeProperties Then _
        members.AddRange(valueType.GetProperties())

      If sett.sortByName Then
        members.Sort(New MemberInfoByNameComparer())
      End If

      For Each fItem As MemberInfo In members

        If TypeOf fItem Is PropertyInfo Then

          Dim pi As PropertyInfo = CType(fItem, PropertyInfo)
          If Not pi.CanRead Then Continue For

          itemValue = valueType.InvokeMember( _
            pi.Name, BindingFlags.GetProperty, Nothing, value, Nothing, Nothing)
        Else
          itemValue = valueType.InvokeMember( _
          fItem.Name, BindingFlags.GetField, Nothing, value, Nothing, Nothing)
        End If

        If itemValue Is Nothing Then
          Dim t As Type
          If TypeOf fItem Is PropertyInfo Then
            t = CType(fItem, PropertyInfo).PropertyType
          Else
            t = CType(fItem, FieldInfo).FieldType
          End If

          pom = CreateTag(fItem.Name, t, True, sett.nullTagItem, sett)
          content.AppendLine(pom)

        Else
          pom = CreateXmlByType(fItem.Name, itemValue, sett)
          content.AppendLine(pom)
        End If

      Next

      Dim ret As String = _
        CreateTag(tag, value.GetType(), False, content.ToString(), sett)

      Return ret

    End Function

    Private Function CreateXmlByType(ByVal tag As String, ByVal value As Object, ByVal sett As Settings) As String

      If sett.maxDepth < sett.currentDepth Then _
        Return ""

      sett.currentDepth += 1

      Dim ret As String

      If TypeOf value Is ValueType Then
        Dim t As Type = value.GetType()
        If t.IsPrimitive Then
          ret = CreateXml(tag, CType(value, ValueType), sett)
        Else
          If sett.expandValueTypes Then
            ret = CreateXml(tag, CType(value, Object), sett)
          Else
            ret = CreateXml(tag, CType(value, ValueType), sett)
          End If
        End If
      ElseIf TypeOf value Is String Then
        ret = CreateXml(tag, CType(value, String), sett)
      ElseIf TypeOf value Is IEnumerable Then
        ret = CreateXml(tag, CType(value, IEnumerable), sett)
      Else
        ret = CreateXml(tag, CType(value, Object), sett)
      End If

      sett.currentDepth -= 1

      Return ret
    End Function

    Private Delegate Sub AddingDelegate(ByVal text As String)
    Private Function CreateTag( _
      ByVal tag As String, ByVal type As Type, ByVal isAtomic As Boolean, ByVal value As String, _
      ByVal sett As Settings) As String

      Dim sb As New StringBuilder
      Dim d As AddingDelegate

      If sett.inlineValues AndAlso isAtomic Then
        d = AddressOf sb.Append
      Else
        d = AddressOf sb.AppendLine
      End If

      If sett.useLongTypeNames Then
        d("<" & tag & " type=""" & type.FullName & """>")
      Else
        d("<" & tag & " type=""" & type.Name & """>")
      End If

      d(value)

      d("</" & tag & ">")

      Return sb.ToString()
    End Function

    Private Function FormatXml(ByVal text As String) As String
      Dim rdr As New System.IO.StringReader(text)
      Dim level As Integer = 0
      Dim ret As New StringBuilder()

      Dim line As String = rdr.ReadLine()

      While line IsNot Nothing
        If line.StartsWith("</") Then
          level = level - 1
          ret.AppendLine(INTENDER.Repeat(level) & line)
        ElseIf line.StartsWith("<") Then
          ret.AppendLine(INTENDER.Repeat(level) & line)
          If (Not line.Contains("</")) Then _
            level = level + 1
        ElseIf line.Trim().Length > 0 Then
          ret.AppendLine(INTENDER.Repeat(level) & line)
        End If

        line = rdr.ReadLine()
      End While

      Return ret.ToString()
    End Function

#End Region 'Private methods

#End Region 'XML



  End Module

End Namespace